### PR TITLE
Fix `NCAMainTest` tests to use labels

### PR DESCRIPTION
**Pull request by rcurtin**
While debugging the test failures in mlpack/ensmallen#201, I found that the reason for these failures was that a few of the NCA tests did not pass in labels.  By default, NCA will use the last dimension of the input as the labels, but for these three test cases, the last dimension only took 0 values.  This in turn led to an objective function gradient of 0 when the NCA optimization began, which in turn meant that the optimization immediately terminated.  For the three tests I have modified here, what is being tested is that two slightly different parameter values give different outputs.

Amusingly, the reason these tests worked before mlpack/ensmallen#201 was not that the results being returned were different---actually they were the same (all NaN, since L-BFGS was forcing a second iteration when the gradient was 0).  But since `NaN != NaN` in C++, it still passed the tests.

I also had to make some robustness changes to the tests; each of these three tests now may require multiple trials, since changing the maximum number of iterations or other optimizer parameters may not always change the results.  I tested each of these tests 1000 times with no failures (against ensmallen modified as per mlpack/ensmallen#201).

So, anyway, in order to merge mlpack/ensmallen#201, we'll first need to merge this. :+1:

### CODE REVIEW

**Comment by zoq** [comment]
Looks like `bigglesworth` is missing curl.

File: src/mlpack/tests/main_tests/nca_test.cpp [hunk 1]
```diff
@@ -383,34 +401,52 @@ BOOST_AUTO_TEST_CASE(NCALinearScanTest2)
...
 BOOST_AUTO_TEST_CASE(NCADifferentNumBasisTest)
 {
-  // Simple dataset.
-  arma::mat x;
-  x.randu(3, 600);
+  // This test can randomly fail and it can be okay, so we run multiple times if
+  // necessary.
+  bool success = false;
+  size_t trial = 0;
+  while (trial < 5)
```
Comment (zoq) [inline src/mlpack/tests/main_tests/nca_test.cpp:408]: I see you used `trial < 3` in the `NCADifferentToleranceTest` test does this one fail more often?

Comment (rcurtin) [reply src/mlpack/tests/main_tests/nca_test.cpp:408]: I'm not sure---I tried 5 trials initially here.  Personally I would believe that this test is more likely to fail more often, but I'm not sure.  I could always bump the `DifferentToleranceTest` to 5 trials; what do you think?

**Comment by rcurtin** [comment]
@mlpack-jenkins test this please

File: src/mlpack/tests/main_tests/nca_test.cpp (hunk 1 above)
Comment (zoq) [reply src/mlpack/tests/main_tests/nca_test.cpp:408]: I think either is fine, was just curious if you have seen a higher fail rate on the test.

Comment (rcurtin) [reply src/mlpack/tests/main_tests/nca_test.cpp:408]: Nope, just an intuitive guess.  Both of the tests did fail if I had no trials though, and the `DifferentToleranceTest` seemed to fail less often (but I didn't write down how often it failed anywhere unfortunately).

**Comment by favre49** [comment]
This is the first time I've merged something with conflicts, hopefully I didn't do something wrong :)

**Comment by favre49** [comment]
I guess I'll wait to see if the tests go green again before merging (it should, but being a bit paranoid)

**Comment by rcurtin** [comment]
@favre49 no worries, it seems like the style checker wants to give me a hard time on this one. :)  I just pushed a few more fixes, let's see if it works this time. :+1:

**Comment by rcurtin** [comment]
Thanks for merging!  Now let's hope that mlpack/ensmallen#201 builds fine. :)
